compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -82,71 +82,423 @@ static void lookupScheme2(TR::Node *node, bool unbalanced, bool fromTableEval, T
+
+   bool is64Bit;
+
+   switch (compareInfo.type)
+      {
+      case TR::Int8:
+      case TR::Int16:
+      case TR::Int32:
+         is64Bit = false;
+         break;
+      case TR::Int64:
+         is64Bit = true;
+         break;
+      case TR::Address:
+         is64Bit = cg->comp()->target().is64Bit();
+         break;
+      default:
+         TR_ASSERT_FATAL_WITH_NODE(node, false, "Cannot call evaluateIntCompareToConditionRegister with data type %s", TR::DataType::getName(compareInfo.type));
+      }
+
+   if (is64Bit && !cg->comp()->target().is64Bit())
+      return evaluateDualIntCompareToConditionRegister(condReg, node, firstChild, secondChild, compareInfo, cg);
+
+   if (is64Bit)
+      {
+      if (compareInfo.isUnsignedOrUnordered)
+         {
+         cmpOp = TR::InstOpCode::cmpl8;
+         cmpiOp = TR::InstOpCode::cmpli8;
+         }
+      else
+         {
+         cmpOp = TR::InstOpCode::cmp8;
+         cmpiOp = TR::InstOpCode::cmpi8;
+         }
+      }
    else
-      return TR::TreeEvaluator::ifiucmpleEvaluator(node, cg);
+      {
+      if (compareInfo.isUnsignedOrUnordered)
+         {
+         cmpOp = TR::InstOpCode::cmpl4;
+         cmpiOp = TR::InstOpCode::cmpli4;
+         }
+      else
+         {
+         cmpOp = TR::InstOpCode::cmp4;
+         cmpiOp = TR::InstOpCode::cmpi4;
+         }
+      }
+
+   TR::Register *firstReg = evaluateAndExtend(firstChild, compareInfo.isUnsignedOrUnordered, false, cg);
+   bool canUseCmpi = secondChild->getOpCode().isLoadConst() &&
+      (compareInfo.isUnsignedOrUnordered
+         ? is16BitUnsigedImmediate(secondChild->get64bitIntegralValueAsUnsigned())
+         : is16BitSignedImmediate(secondChild->get64bitIntegralValue()));
+
+   if (canUseCmpi)
+      {
+      generateTrg1Src1ImmInstruction(cg, cmpiOp, node, condReg, firstReg, secondChild->get64bitIntegralValue());
+      }
+   else
+      {
+      TR::Register *secondReg = evaluateAndExtend(secondChild, compareInfo.isUnsignedOrUnordered, false, cg);
+      generateTrg1Src2Instruction(cg, cmpOp, node, condReg, firstReg, secondReg);
+      stopUsingExtendedRegister(secondReg, secondChild, cg);
+      }
+
+   stopUsingExtendedRegister(firstReg, firstChild, cg);
+
+   return compareInfo.cond;
    }
 
-TR::Register *OMR::Power::TreeEvaluator::acmpltEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+CompareCondition evaluateFloatCompareToConditionRegister(
+   TR::Register *condReg,
+   TR::Node *node,
+   TR::Node *firstChild,
+   TR::Node *secondChild,
+   const CompareInfo& compareInfo,
+   TR::CodeGenerator *cg
+)
+   {
+   CRCompareCondition crCond = compareConditionInCR(compareInfo.cond);
+   TR::Register *firstReg = cg->evaluate(firstChild);
+   TR::Register *secondReg = cg->evaluate(secondChild);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::fcmpu, node, condReg, firstReg, secondReg);
+   if (crCond.isReversed != compareInfo.isUnsignedOrUnordered)
+      generateTrg1Src2ImmInstruction(
```

Comment:
I think you've misunderstood something here. The purpose of the `isUnsignedOrUnordered` flag in the floating-point case is to determine whether or not unordered operands should return true for the comparison, as we have both variants for every comparison and those can be used internally by the optimizer if the result of a floating-point comparison was negated, e.g. `!(x < y) -> fcmpgeu`. For that reason, checking that flag in addition to whether the CCR bit being checked is reversed is critical to correctness.

I agree that the name of the flag is a bit iffy, though. The same POD type is used to represent both integral and floating-point comparisons in a unified manner, as this will be used in a future PR to build some nicer abstractions that can allow for non-comparison evaluators to evaluate a child node to a CCR, e.g. for `iselect`. I'd be open to splitting this into 2 flags if you think it'd make this more clear, although that would mean having flags in the struct that don't make sense for certain comparisons (what would it mean to compare unsigned floats?).

Reply:
I would suggest naming it something like:  isUnsignedIntOrFPCaresUnordered,  making clear this field's intention.
Making it into two separate fields is fine too, commenting one for integral types and another for FP.

Reply:
I feel that a name like that would be getting a bit too long and would just be ignoring the real understandability problem here. I don't think keeping these flags together actually provides much value thinking back, as we never construct `CompareInfo` objects in large numbers. I've gone ahead and split this into two separate flags to make things easier to understand.